tool/release/CreateNotes.kt

```diff
@@ -0,0 +1,216 @@
+            .buildGetRequest(GenericUrl("https://api.github.com/repos/$org/$repo/commits/$commit/pulls"))
+            .setHeaders(
+                HttpHeaders().setAuthorization("Token $githubToken").setAccept("application/vnd.github.v3+json")
+            )
+            .execute()
+        val body = Json.parse(String(response.content.readBytes()))
+        val prs = body.asArray()
+        if (prs.size() > 0) {
+            val notes = prs.map { pr ->
+                val types = pr.asObject().get("labels").asArray().map { e -> e.asObject().get("name").asString() }
+                    .filter { e -> e.startsWith("type:") }
+                val type =
+                    if (types.contains("type: feature")) CommitNoteType.FEATURE
+                    else if (types.contains("type: bug")) CommitNoteType.BUG
+                    else if (types.contains("type: refactor")) CommitNoteType.REFACTOR
+                    else CommitNoteType.OTHER
+                CommitNote(
+                    title = pr.asObject().get("title").asString(),
+                    desc = pr.asObject().get("body").asString(),
+                    type = type
+                )
+            }
+            notes
+        } else {
+            val response = NetHttpTransport()
+                .createRequestFactory()
+                .buildGetRequest(GenericUrl("https://api.github.com/repos/$org/$repo/commits/$commit"))
+                .setHeaders(
+                    HttpHeaders().setAuthorization("Token $githubToken").setAccept("application/vnd.github.v3+json")
+                )
+                .execute()
+            val body = Json.parse(String(response.content.readBytes()))
+            val notes = listOf(
+                CommitNote(
+                    title = body.asObject().get("commit").asObject().get("message").asString(),
+                    desc = "",
+                    type = CommitNoteType.OTHER
+                )
+            )
+            notes
+        }
+    }
+}
+
+fun createCommitNote(note: CommitNote): String {
+    var desc = StringBuilder()
+    var header = 0
+    for (line in note.desc.lines()) {
+        if (line.startsWith("##")) {
+            header += 1
+        } else if (header == 1) {
+            desc.append(line)
+        } else if (header > 1) {
+            break
+        }
+    }
+    return "- **${note.title}**\n  $desc"
+}
+
+fun createReleaseNote(commitNotes: List<CommitNote>, releaseTemplateFile: Path) {
+    val features = mutableListOf<CommitNote>()
+    val bugs = mutableListOf<CommitNote>()
+    val refactors = mutableListOf<CommitNote>()
+    val others = mutableListOf<CommitNote>()
+
+    commitNotes.forEach { note ->
+        when (note.type) {
+            CommitNoteType.FEATURE -> features.add(note)
+            CommitNoteType.BUG -> bugs.add(note)
+            CommitNoteType.REFACTOR -> refactors.add(note)
+            else -> others.add(note)
+        }
+    }
+
+    val releaseNote = """
```

Comment:
@lolski could use better-looking indentation if you call `.trimIndent()` on it:

```
val releaseNote = """
        Install & Run: http://docs.vaticle.com/docs/running-typedb/install-and-run
        ## New Features
        ${features.map { feature -> createCommitNote(feature) }.joinToString("\n")}

        ## Bugs Fixed
        ${bugs.map { bug -> createCommitNote(bug) }.joinToString("\n")}

        ## Code Refactor
        ${refactors.map { refactor -> createCommitNote(refactor) }.joinToString("\n")}

        ## Other Improvements
        ${others.map { other -> createCommitNote(other) }.joinToString("\n")}
""".trimIndent()
Files.write(releaseTemplateFile, releaseNote.toByteArray(UTF_8))
```

Reply:
I tried that but it's not working properly.

This format
```
    val markdown = """
        Install & Run: $installInstruction
        
        ## New Features
        ${features.map(::createCommitNoteMd).joinToString("\n")}
        ...
    """.trimIndent()
```

resulted in an overindented first few lines:

```
        Install & Run: http://docs.vaticle.com/docs/running-typedb/install-and-run
        
        ## New Features
        - **Implement rule negation validation to prevent contradictions**
  To properly support negation in rules, we must validate that rules cannot contradict themselves and reach a stable final solution, we require that rules containing negations do not recurse through the negation - this is called stratified negation.This PR implements the commit time cycle detection required to ban cyclical rules through negations.
- **Optimise resolution of fully bound concludables**
  In some cases, a concludable that is being resolved may have all of its retrievable variables (named and anonymous) fully bound. This means that only a single answer is required to prove the concludable holds, rather than all possible ways. We add short circuiting to prevent exploring further rules in a concludable when a single answer has been found.
```